OvmfPkg/Library/PlatformInitLib/MemDetect.c

```diff
@@ -898,6 +898,69 @@ PlatformScanHostProvided64BitPciMmioEnd (
   return EFI_NOT_FOUND;
 }
 
+VOID
+EFIAPI
+Switch4Level (
+  VOID
+  );
+
+STATIC
+VOID
+PlatformSetupPagingLevel (
+  IN OUT EFI_HOB_PLATFORM_INFO  *PlatformInfoHob
+  )
+{
+ #ifdef MDE_CPU_X64
+  UINT32      PagingLevel;
+  EFI_STATUS  Status;
+  IA32_CR4    Cr4;
+
+  Cr4.UintN = AsmReadCr4 ();
+  if (!Cr4.Bits.LA57) {
+    DEBUG ((DEBUG_INFO, "%a: using 4-level paging (la57 not supported)\n", __func__));
+    return;
+  }
+
+  Status = QemuFwCfgParseUint32 (
+             "opt/org.tianocode/PagingLevel",
+             FALSE,
+             &PagingLevel
+             );
+  switch (Status) {
+    case EFI_NOT_FOUND:
+      if (PlatformInfoHob->FirstNonAddress < (1ll << 40)) {
+        //
+        // If the highest address actually used is below 1TB switch back into
+        // 4-level paging mode for better compatibility with older guests.
+        //
+        DEBUG ((DEBUG_INFO, "%a: using 4-level paging (default for small guest)\n", __func__));
+        PagingLevel = 4;

```

Comment:
I am confused on the what is final decision for 5-level or 4-level. Do we have a table to summarize?

If we prefer such rule, why not create 4-level in the beginning?
There is no need to create 5-level in the beginning then switch back to 4-level later.

Reply:
> I am confused on the what is final decision for 5-level or 4-level. Do we have a table to summarize?

In case `opt/org.tianocore/PagingLevel` fw_cfg file exists use whatever it says.

Otherwise: Large guests (more than 1TB address space needed for main memory and possible additional reservations configured via fw_cfg) will use 5-level paging, otherwise 4-level paging is used.

> If we prefer such rule, why not create 4-level in the beginning? There is no need to create 5-level in the beginning then switch back to 4-level later.

The initial page tables are created by reset vector, and it simply uses 5-level paging if la57 and gigabyte pages are supported by the processor.  At that point the data used by this patch to decide whenever 4-level or 5-level paging should be used is not available yet.

Going from 5-level to 4-level paging is easier than the other way around, you just have to dereference first entry in the 5th paging level.

Reply:
How do you plan to use `PcdUse5LevelPageTable`, https://github.com/tianocore/edk2/blob/master/MdeModulePkg/MdeModulePkg.dec#L2165-L2170

Reply:
With `PcdUse5LevelPageTable=FALSE` the OVMF ResetVector code will never turn on 5-level paging, so OVMF will run in 4-level paging mode no matter what.

It probably makes sense to check `PcdUse5LevelPageTable` for logging, it is a difference whenever we run in 4-level paging mode due to the /build/ not supporting 5-level paging or due to the /cpu/ not supporting 5-level paging.

Reply:
So far, I can see:
1) EDKII: PcdUse5LevelPageTable
2) QEMU: opt/org.tianocore/PagingLevel
3) CPU: MAX CPU Address supported by CPU
4) PI: CPU_HOB.SizeOfMemorySpace (https://uefi.org/specs/PI/1.8A/V3_HOB_Code_Definitions.html#efi-hob-cpu) (MAX Address supported by Memory Controller)
5) Platform: PlatformInfoHob->FirstNonAddress (MAX DRAM + MMIO Address on the platform)

Can we add a table to describe the rule?

Reply:
Added longish comment to `PlatformSetupPagingLevel()` describing behavior for both ResetVector and that function.

Reply:
Pushed update with comment clarified.

Reply:
Thanks. Additional question.

It seems to me that only `name=opt/org.tianocode/PagingLevel,string=4` is useful, which is used to control the legacy behavior.

What is the difference between `name=opt/org.tianocode/PagingLevel,string=5` and name being absent? If they are same, why 5 is needed?

Reply:
By default OVMF will downgrade to 4-level paging for small (less than 1TB address space needed) guests.  The fw_cfg option allows to override that heuristic both ways, i.e. force 4-level paging for large guests and force 5-level paging for small guests.

Reply:
The force code path is: QemuFwCfgParseUint32() -> case EFI_SUCCESS -> "fw_cfg override" message.

If PagingLevel == 4 -> code will downgrade to 4-level mode via Switch4Level()
If PagingLevel == 5 -> code will continue in 5-level mode.

Reply:
I still dont understand what is the difference between "PagingLevel == 5" and no PagingLevel parameter.

I feel they are same, just continue.

So, if they are same, then why we need to define PagingLevel == 5.

Reply:
It is hard for me to follow.

From your patch code, I dont see you do anything for PageLevel=5.
Where is your code to enforce `PagingLevel=5 -> use 5-level paging` ?

Reply:
Turning in 5-level paging happens in the OVMF reset vector when supported by the CPU.  So there is no need to handle the 4 -> 5 switch here.  If we land in this function with 4-level paging mode being active the CPU does not support 5-level paging (or `PcdUse5LevelPageTable` is `FALSE`) and we have no choice anyway.  So:

* if current paging level is 4
   * continue to run with 4-level paging no matter what.
* if current paging level is 5
   * PagingLevel=4 -> use level 4
   * PagingLevel=5 -> use level 5
   * no PagingLevel -> choose paging level depending on guest RAM size
     * less than 1TB of memoy -> use level 4
     * more than 1TB of memory -> use level 5

Reply:
OK, I got your point finally.

I think I am mislead by the line 983, `PagingLevel = 5`. Let me put a comment there.